rfcs/20200422-concurrent-global-scratchpad-mode/README.md

```diff
@@ -0,0 +1,185 @@
+# Proposal for introducing concurrent and global scratchpad mode
+
+## Background
+
+Currently, some primitive implementations in oneDNN require a temporary storage
+for computations, a.k.a. scratchpad, and the library supports two scratchpad
+policies: global and concurrent.
+
+In most cases, the scratchpad has a concurrent mode, whereby each thread
+allocates a new scratchpad storage during the instantiation of a primitive that
+has been defined without `USE_GLOBAL_SCRATCHPAD` flag in the library. The
+following is an example of an implementation definition with concurrent
+scratchpad:
+
+~~~cpp
+
+DECLARE_COMMON_PD_T("ncsp_bnorm:any", ncsp_batch_normalization_fwd_t);
+
+~~~
+
+In case a primitive implementation does contain the `USE_GLOBAL_SCRATCHPAD`
+flag, it can create (and expand, if necessary) a global scratchpad that is
+shared among all primitive implementations defined with this flag. Such
+scratchpad is defined as `static`, and it is destroyed once the last primitive
+in user's application is also destroyed. Since the global scratchpad is defined
+with a `thread_local` qualifier, the user has to ensure that all primitives that
+use such storage are created **and** executed on the same thread. The default
+library build configuration enables global scratchpad support, and it can be
+disabled using the `DNNL_ENABLE_CONCURRENT_EXEC` macro. The following is an
+example of an implementation definition with global scratchpad support:
+
+~~~cpp
+DECLARE_COMMON_PD_T(
+                GEMM_IMPL_STR, gemm_convolution_fwd_t, USE_GLOBAL_SCRATCHPAD);
+~~~
+
+Both types of scratchpad are internal to the library, and currently there are
+two ways a user can have a control over the scratchpad:
+
+1. By providing primitive attributes with a `scratchpad_mode::user` flag.
+
+2. By querying the scratchpad size the primitive implementation requires using
+   `query::memory_consumption_s64`.
+
+In case of 1, the library does not allocate an internal storage and relies on
+the user to provide one during execution using `DNNL_ARG_SCRATCHPAD`. Although
+this ensures thread safety by providing independent (i.e. concurrent) storage
+for every primitive, this can dramatically increase memory consumption by all
+scratchpad storages allocated by the user.
+
+In case of 2, the library will either allocate an internal buffer of such size,
+or expects the user to provide one during execution. In case the library
+allocates an internal scratchpad, the user has no knowledge whether it is shared
+among all or some primitives or not. Therefore, the query cannot guarantee the
+total memory consumption if executing more than one primitive.
+
+
+## Objective
+
+The aim of this proposal is to more provide more explicit controls to the user
+over primitive scratchpad. The main goal is to reduce the total memory usage due
+to scratchpad and/or guarrantee thread safety by using (i) new scratchpad modes
+and (ii) new compilation flag.
```

Comment:
I'm actually concerned with the flexibility we are going to provide, not the number of options but the limitations that each of them has. It was challenging for users to deal with `user` and `library` modes until @mgouicem documented all the limitations but now we have two more modes. Although the API is common for CPU and GPU the behavior of some policies can potentially differ. For instance, we don't support global scratchpad for GPU so far, but if we decide to implement it there will be no requirement to create and run a primitive in the same thread because there is no sense in making scratchpad tread local for GPU (unless I'm missing something). That means that CPU and GPU primitives that use `thread_private_scratchpad_t` will behave differently and we need to document it.

The ability to create a primitive in one thread but execute it in another was probably useful for frameworks because they could use it to implement their own primitive cache.

Given that the library provides the primitive cache and that it will always be possible to create and run a GPU primitive in different threads, what if we make the `library` mode to always use `thread_private_scratchpad_t` for CPU and `primitive_private_scratchpad_t` for GPU and provide a compile time flag for the case if somebody really needs to use only `primitive_private_scratchpad_t`?
This `library` mode will potentially have the difference in behavior of  CPU and GPU primitive created using the same but we do not introduce new policies.

The general question would be, how important is it to expose all the suggested policies to user? In other words, do we have a clear understanding of the real use cases?

Reply:
Thank you very much for the details.

As it is hopefully described in the background of the RFC, the main motivation came from the fact that only a fraction of primitives on the CPU are currently using a global scratchpad: while user can save memory when using these on a topology level, the property of certain primitives to have a global scratchpad is not transparent for the user (unless you look into the code). This makes the scratchpad policy inconsistent throughout the library, therefore the first idea was to have a toggle for all primitives on CPU. Now the compile flag becomes the only way to manage scratchpad policy, therefore an API would take care of this issue at run-time.

Perhaps, as you suggest, we can simply change the default for CPU primitives to use `thread_private` and GPU to use `primitive_private` (probably also change `thread_private` to `primitive_shared` to account for potentially introducing a global scratchpad on GPU), and tweak it with a compile flag.

The immediate use case would be for frameworks that ensure creating and using CPU primitives on the same thread, therefore `primitive_shared`/`thread_private` would be beneficial for them.

Reply:
I understand that the main benefit would be a decreased memory consumption. But given that some frameworks already use the `library` mode, that as you mentioned may or may not use the global scratchpad, they are fine with the requirement that the global scratchpad has. So if we switch to the global scratchpad for all CPU primitives then the frameworks will benefit from it without any new policies and changes in their code.
But those frameworks that are not fine with the requirement can just continue using the compile time flag that we already provide to switch to the so called concurrent scratchpad.
To sum it up, with this approach the behavior of the `library` mode stays aligned with the documentation while making it more efficient in terms of memory consumption.

Reply:
I agree with your idea. In this case I will remove references of new API from this RFC, and will limit the scope to adding support for global scratchpad for all CPU primitives.